src/Agent.PluginCore/PluginUtil.cs

```diff
@@ -0,0 +1,1611 @@
+
+                    if (signaled == outputSignal)
+                    {
+                        ProcessOutput();
+                    }
+                    else
+                    {
+                        _stopWatch.Stop();
+                        break;
+                    }
+                }
+
+                // Just in case there was some pending output when the process shut down go ahead and check the
+                // data buffers one last time before returning
+                ProcessOutput();
+
+                executionContext.Debug($"Finished process with exit code {_proc.ExitCode}, and elapsed time {_stopWatch.Elapsed}.");
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Wait for process to finish.
+            if (_proc.ExitCode != 0 && requireExitCodeZero)
+            {
+                throw new ProcessExitCodeException(exitCode: _proc.ExitCode, fileName: fileName, arguments: arguments);
+            }
+
+            return _proc.ExitCode;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_proc != null)
+                {
+                    _proc.Dispose();
+                    _proc = null;
+                }
+            }
+        }
+
+        private void ProcessOutput()
+        {
+            List<string> errorData = new List<string>();
+            List<string> outputData = new List<string>();
+
+            string errorLine;
+            while (_errorData.TryDequeue(out errorLine))
+            {
+                errorData.Add(errorLine);
+            }
+
+            string outputLine;
+            while (_outputData.TryDequeue(out outputLine))
+            {
+                outputData.Add(outputLine);
+            }
+
+            _outputProcessEvent.Reset();
+
+            // Write the error lines.
+            if (errorData != null && this.ErrorDataReceived != null)
+            {
+                foreach (string line in errorData)
+                {
+                    if (line != null)
+                    {
+                        this.ErrorDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
+
+            // Process the output lines.
+            if (outputData != null && this.OutputDataReceived != null)
+            {
+                foreach (string line in outputData)
+                {
+                    if (line != null)
+                    {
+                        // The line is output from the process that was invoked.
+                        this.OutputDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
+        }
+
+        private async Task CancelAndKillProcessTree()
```

Comment:
i'm curious how cancellation works when initiated from the agent. let's discuss offline

Reply:
decrease cancel timeout 5 secs.

Reply:
fixed.